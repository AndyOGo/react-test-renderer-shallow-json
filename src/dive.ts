import type { ReactTestRendererTreeNodeFixed } from './types';

export function dive(
  tree: ReactTestRendererTreeNodeFixed,
  depth: number
): number {
  return typeof tree === 'string' || tree.nodeType === 'host'
    ? depth
    : depth - 1;
}
