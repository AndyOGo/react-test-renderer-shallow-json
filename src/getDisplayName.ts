import {
  ReactTestRendererTreeFixed,
  isContext,
  isNamedExoticComponentType,
} from './types';

const elementSymbolMap: Readonly<Record<symbol, string>> = {
  [Symbol.for('react.memo')]: 'Memo',
  [Symbol.for('react.context')]: 'Consumer',
  [Symbol.for('react.provider')]: 'Provider',
};

function appendSymbol(displayName: string, $$typeof?: symbol): string {
  if (!$$typeof || !elementSymbolMap[$$typeof]) {
    return displayName;
  }

  return `${displayName}.${elementSymbolMap[$$typeof]}`;
}

export function getDisplayName(
  type: ReactTestRendererTreeFixed['type'],
  fallback = '???'
): string {
  if (typeof type === 'string') {
    return type;
  }

  if (isContext(type)) {
    return appendSymbol(type._context.displayName || fallback, type.$$typeof);
  }

  if (isNamedExoticComponentType(type)) {
    return (
      type.displayName ||
      appendSymbol(
        type.type ? getDisplayName(type.type) : fallback,
        type.$$typeof
      )
    );
  }

  return type.displayName || type.name || fallback;
}
