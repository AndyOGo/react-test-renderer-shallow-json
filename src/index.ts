console.log('Start coding here...');
