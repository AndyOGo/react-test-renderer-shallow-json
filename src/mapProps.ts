import type { ReactTestRendererTreeFixed } from './types';

export function mapProps(props: ReactTestRendererTreeFixed['props']) {
  const { children, ...propsWithoutChildren } = props;

  return propsWithoutChildren;
}
