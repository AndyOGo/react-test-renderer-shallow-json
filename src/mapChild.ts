import type {
  ReactTestRendererNodeFixed,
  ReactTestRendererTreeChild,
} from './types';
import { getDisplayName } from './getDisplayName';
import { mapProps } from './mapProps';
import { mapChildren } from './mapChildren';

export function mapChild(
  child: ReactTestRendererTreeChild
): ReactTestRendererNodeFixed {
  if (typeof child === 'string') {
    return child;
  }

  const { key, ref } = child;

  const json = {
    type: getDisplayName(child.type),
    props: mapProps({
      ...child.props,
      ...(key && { key }),
      ...(ref && { ref }),
    }),
    children: mapChildren(child.props),
  };

  Object.defineProperty(json, '$$typeof', {
    value: Symbol.for('react.test.json'),
  });

  return json;
}
