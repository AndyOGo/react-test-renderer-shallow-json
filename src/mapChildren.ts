import type {
  ReactTestRendererJSONFixed,
  ReactTestRendererTreeFixed,
} from './types';
import { mapChild } from './mapChild';

export function mapChildren(
  props: ReactTestRendererTreeFixed['props']
): ReactTestRendererJSONFixed['children'] {
  const { children } = props;

  if (!children) {
    return null;
  }

  if (!Array.isArray(children)) {
    return [mapChild(children)];
  }

  return children.map(mapChild);
}
