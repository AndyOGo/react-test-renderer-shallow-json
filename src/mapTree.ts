import type { ReactTestRendererNode } from 'react-test-renderer';
import type { ReactTestRendererTreeNodeFixed } from './types';
import { getDisplayName } from './getDisplayName';
import { mapProps } from './mapProps';
import { mapChildren } from './mapChildren';
import { mapTrees } from './mapTrees';
import { withReactTestSymbol } from './withReactTestSymbol';

export function mapTree(
  tree: ReactTestRendererTreeNodeFixed,
  depth = 1
): ReactTestRendererNode {
  if (typeof tree === 'string') {
    return tree;
  }

  const json = {
    type: getDisplayName(tree.type),
    props: mapProps(tree.props),
    children:
      depth > 0 ? mapTrees(tree.rendered, depth) : mapChildren(tree.props),
  };

  return withReactTestSymbol(json) as ReactTestRendererNode;
}
