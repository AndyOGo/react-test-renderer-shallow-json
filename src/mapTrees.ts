import type { ReactTestRendererJSON } from 'react-test-renderer';
import type { ReactTestRendererTreeFixed } from './types';
import { mapTree } from './mapTree';
import { dive } from './dive';

export function mapTrees(
  trees: ReactTestRendererTreeFixed['rendered'],
  level = 1
): ReactTestRendererJSON['children'] {
  if (!trees) {
    return null;
  }

  if (Array.isArray(trees)) {
    return trees.map((tree) => mapTree(tree, dive(tree, level)));
  }

  return [mapTree(trees, dive(trees, level))];
}
