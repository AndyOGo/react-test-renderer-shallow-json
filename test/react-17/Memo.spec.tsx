import React from 'react';
import { create } from 'react-test-renderer';
import { shallowJSON } from '../../src';
import {
  Empty,
  Button,
  ListItem,
  List,
  TableHead,
  TableHeadCell,
  Table,
  TableRow,
  TableFoot,
  TableCell,
  TableBody,
  EmptyNamed,
  ButtonNamed,
  ListItemNamed,
  ListNamed,
  TableHeadNamed,
  TableHeadCellNamed,
  TableNamed,
  TableRowNamed,
  TableFootNamed,
  TableCellNamed,
  TableBodyNamed,
} from './Memo';

describe('Empty', () => {
  it('matches snapshot', () => {
    expect(shallowJSON(create(<Empty />).toTree())).toMatchSnapshot();
  });

  it('matches named snapshot', () => {
    expect(shallowJSON(create(<EmptyNamed />).toTree())).toMatchSnapshot();
  });
});

describe('Button', () => {
  it('matches snapshot', () => {
    expect(
      shallowJSON(create(<Button onClick={jest.fn()} />).toTree())
    ).toMatchSnapshot();
  });

  it('matches named snapshot', () => {
    expect(
      shallowJSON(create(<ButtonNamed onClick={jest.fn()} />).toTree())
    ).toMatchSnapshot();
  });
});

describe('List', () => {
  it('matches snapshot', () => {
    expect(
      shallowJSON(
        create(
          <List>
            {['A', 'B', 'C'].map((children, index) => (
              <ListItem key={index}>{children}</ListItem>
            ))}
          </List>
        ).toTree()
      )
    ).toMatchSnapshot();
  });

  it('matches named snapshot', () => {
    expect(
      shallowJSON(
        create(
          <ListNamed>
            {['A', 'B', 'C'].map((children, index) => (
              <ListItemNamed key={index}>{children}</ListItemNamed>
            ))}
          </ListNamed>
        ).toTree()
      )
    ).toMatchSnapshot();
  });
});

describe('Table', () => {
  it('matches snapshot', () => {
    expect(
      shallowJSON(
        create(
          <Table>
            <TableHead>
              <TableRow>
                <TableHeadCell>Head A</TableHeadCell>
                <TableHeadCell>Head B</TableHeadCell>
                <TableHeadCell>Head C</TableHeadCell>
              </TableRow>
            </TableHead>

            <TableFoot>
              <TableRow>
                <TableCell>Foot A</TableCell>
                <TableCell>Foot B</TableCell>
                <TableCell>Foot C</TableCell>
              </TableRow>
            </TableFoot>

            <TableBody>
              {[
                ['A', 'B', 'C'],
                ['X', 'Y', 'Z'],
              ].map((row, rowIndex) => (
                <TableRow key={rowIndex}>
                  {row.map((cell, cellIndex) => (
                    <TableCell key={cellIndex}>{cell}</TableCell>
                  ))}
                </TableRow>
              ))}
            </TableBody>
          </Table>
        ).toTree()
      )
    ).toMatchSnapshot();
  });

  it('matches named snapshot', () => {
    expect(
      shallowJSON(
        create(
          <TableNamed>
            <TableHeadNamed>
              <TableRowNamed>
                <TableHeadCellNamed>Head A</TableHeadCellNamed>
                <TableHeadCellNamed>Head B</TableHeadCellNamed>
                <TableHeadCellNamed>Head C</TableHeadCellNamed>
              </TableRowNamed>
            </TableHeadNamed>

            <TableFootNamed>
              <TableRowNamed>
                <TableCellNamed>Foot A</TableCellNamed>
                <TableCellNamed>Foot B</TableCellNamed>
                <TableCellNamed>Foot C</TableCellNamed>
              </TableRowNamed>
            </TableFootNamed>

            <TableBodyNamed>
              {[
                ['A', 'B', 'C'],
                ['X', 'Y', 'Z'],
              ].map((row, rowIndex) => (
                <TableRowNamed key={rowIndex}>
                  {row.map((cell, cellIndex) => (
                    <TableCellNamed key={cellIndex}>{cell}</TableCellNamed>
                  ))}
                </TableRowNamed>
              ))}
            </TableBodyNamed>
          </TableNamed>
        ).toTree()
      )
    ).toMatchSnapshot();
  });
});
