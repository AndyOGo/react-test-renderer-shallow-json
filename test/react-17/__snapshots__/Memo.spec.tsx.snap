// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches named snapshot 1`] = `
<ButtonDisplayName
  onClick={[MockFunction]}
>
  <button
    onClick={[MockFunction]}
    type="button"
  />
</ButtonDisplayName>
`;

exports[`Button matches snapshot 1`] = `
<ButtonDisplayName
  onClick={[MockFunction]}
>
  <button
    onClick={[MockFunction]}
    type="button"
  />
</ButtonDisplayName>
`;

exports[`Empty matches named snapshot 1`] = `<EmptyDisplayName />`;

exports[`Empty matches snapshot 1`] = `<EmptyDisplayName />`;

exports[`List matches named snapshot 1`] = `
<ListDisplayName>
  <ul>
    <ListItemDisplayName
      key="0"
    >
      A
    </ListItemDisplayName>
    <ListItemDisplayName
      key="1"
    >
      B
    </ListItemDisplayName>
    <ListItemDisplayName
      key="2"
    >
      C
    </ListItemDisplayName>
  </ul>
</ListDisplayName>
`;

exports[`List matches snapshot 1`] = `
<ListDisplayName>
  <ul>
    <MemoListItemDisplayName
      key="0"
    >
      A
    </MemoListItemDisplayName>
    <MemoListItemDisplayName
      key="1"
    >
      B
    </MemoListItemDisplayName>
    <MemoListItemDisplayName
      key="2"
    >
      C
    </MemoListItemDisplayName>
  </ul>
</ListDisplayName>
`;

exports[`Table matches named snapshot 1`] = `
<TableDisplayName>
  <table>
    <TableHeadDisplayName>
      <TableRowDisplayName>
        <TableHeadCellDisplayName>
          Head A
        </TableHeadCellDisplayName>
        <TableHeadCellDisplayName>
          Head B
        </TableHeadCellDisplayName>
        <TableHeadCellDisplayName>
          Head C
        </TableHeadCellDisplayName>
      </TableRowDisplayName>
    </TableHeadDisplayName>
    <TableFootDisplayName>
      <TableRowDisplayName>
        <TableCellDisplayName>
          Foot A
        </TableCellDisplayName>
        <TableCellDisplayName>
          Foot B
        </TableCellDisplayName>
        <TableCellDisplayName>
          Foot C
        </TableCellDisplayName>
      </TableRowDisplayName>
    </TableFootDisplayName>
    <TableBodyDisplayName>
      <TableRowDisplayName
        key="0"
      >
        <TableCellDisplayName
          key="0"
        >
          A
        </TableCellDisplayName>
        <TableCellDisplayName
          key="1"
        >
          B
        </TableCellDisplayName>
        <TableCellDisplayName
          key="2"
        >
          C
        </TableCellDisplayName>
      </TableRowDisplayName>
      <TableRowDisplayName
        key="1"
      >
        <TableCellDisplayName
          key="0"
        >
          X
        </TableCellDisplayName>
        <TableCellDisplayName
          key="1"
        >
          Y
        </TableCellDisplayName>
        <TableCellDisplayName
          key="2"
        >
          Z
        </TableCellDisplayName>
      </TableRowDisplayName>
    </TableBodyDisplayName>
  </table>
</TableDisplayName>
`;

exports[`Table matches snapshot 1`] = `
<TableDisplayName>
  <table>
    <MemoTableHeadDisplayName>
      <MemoTableRowDisplayName>
        <MemoTableHeadCellDisplayName>
          Head A
        </MemoTableHeadCellDisplayName>
        <MemoTableHeadCellDisplayName>
          Head B
        </MemoTableHeadCellDisplayName>
        <MemoTableHeadCellDisplayName>
          Head C
        </MemoTableHeadCellDisplayName>
      </MemoTableRowDisplayName>
    </MemoTableHeadDisplayName>
    <MemoTableFootDisplayName>
      <MemoTableRowDisplayName>
        <MemoTableCellDisplayName>
          Foot A
        </MemoTableCellDisplayName>
        <MemoTableCellDisplayName>
          Foot B
        </MemoTableCellDisplayName>
        <MemoTableCellDisplayName>
          Foot C
        </MemoTableCellDisplayName>
      </MemoTableRowDisplayName>
    </MemoTableFootDisplayName>
    <MemoTableBodyDisplayName>
      <MemoTableRowDisplayName
        key="0"
      >
        <MemoTableCellDisplayName
          key="0"
        >
          A
        </MemoTableCellDisplayName>
        <MemoTableCellDisplayName
          key="1"
        >
          B
        </MemoTableCellDisplayName>
        <MemoTableCellDisplayName
          key="2"
        >
          C
        </MemoTableCellDisplayName>
      </MemoTableRowDisplayName>
      <MemoTableRowDisplayName
        key="1"
      >
        <MemoTableCellDisplayName
          key="0"
        >
          X
        </MemoTableCellDisplayName>
        <MemoTableCellDisplayName
          key="1"
        >
          Y
        </MemoTableCellDisplayName>
        <MemoTableCellDisplayName
          key="2"
        >
          Z
        </MemoTableCellDisplayName>
      </MemoTableRowDisplayName>
    </MemoTableBodyDisplayName>
  </table>
</TableDisplayName>
`;
